pkg/scrapers/twitter/auth.go

```diff
@@ -2,35 +2,98 @@ package twitter
 	twitterscraper "github.com/masa-finance/masa-twitter-scraper"
+	"github.com/sirupsen/logrus"
 )
 
-// Login attempts to log in to the Twitter scraper service.
-// It supports three modes of operation:
-// 1. Basic login using just a username and password.
-// 2. Login requiring an email confirmation, using a username, password, and email address.
-// 3. Login with two-factor authentication, using a username, password, and 2FA code.
-// Parameters:
-//   - scraper: A pointer to an instance of the twitterscraper.Scraper.
-//   - credentials: A variadic list of strings representing login credentials.
-//     The function expects either two strings (username, password) for basic login,
-//     or three strings (username, password, email/2FA code) for email confirmation or 2FA.
-//
-// Returns an error if login fails or if an invalid number of credentials is provided.
+type TwitterAccount struct {
+	Username         string
+	Password         string
+	TwoFACode        string
+	RateLimitedUntil time.Time
+}
+
+type TwitterAccountManager struct {
+	accounts []*TwitterAccount
+	index    int
+	mutex    sync.Mutex
+}
+
+func NewTwitterAccountManager(accounts []*TwitterAccount) *TwitterAccountManager {
+	return &TwitterAccountManager{
+		accounts: accounts,
+		index:    0,
+	}
+}
+
+func (manager *TwitterAccountManager) GetNextAccount() *TwitterAccount {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	for i := 0; i < len(manager.accounts); i++ {
+		account := manager.accounts[manager.index]
+		manager.index = (manager.index + 1) % len(manager.accounts)
+		if time.Now().After(account.RateLimitedUntil) {
+			return account
+		}
+	}
+	return nil
+}
+
+func (manager *TwitterAccountManager) MarkAccountRateLimited(account *TwitterAccount) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	account.RateLimitedUntil = time.Now().Add(time.Hour)
+}
+
+func Auth(account *TwitterAccount) *twitterscraper.Scraper {
+	scraper := twitterscraper.New()
+	baseDir := config.GetInstance().MasaDir
+
+	if err := LoadCookies(scraper, account, baseDir); err == nil {
+		logrus.Debugf("Cookies loaded for user %s.", account.Username)
+		if IsLoggedIn(scraper) {
+			logrus.Debugf("Already logged in as %s.", account.Username)
+			return scraper
+		}
+	}
+
+	time.Sleep(100 * time.Millisecond)
+
+	var err error
+	if account.TwoFACode != "" {
+		err = Login(scraper, account.Username, account.Password, account.TwoFACode)
+	} else {
+		err = Login(scraper, account.Username, account.Password)
+	}
+
+	if err != nil {
+		logrus.WithError(err).Warnf("Login failed for %s", account.Username)
+		return nil
+	}
+
+	time.Sleep(100 * time.Millisecond)
+
+	if err = SaveCookies(scraper, account, baseDir); err != nil {
+		logrus.WithError(err).Errorf("Failed to save cookies for %s", account.Username)
+	}
+
+	logrus.Debugf("Login successful for %s", account.Username)
+	return scraper
+}
+
 func Login(scraper *twitterscraper.Scraper, credentials ...string) error {
```

Comment:
I know it's not part of the changeset, however - as you are with this in refactoring: it would make sense for these methods to actually be part of `scraper *twitterscraper.Scraper`:

```golang
func (scraper *twitterscraper.Scraper) Login( credentials ...string) error {
```

Reply:
https://github.com/masa-finance/masa-oracle/pull/576/commits/e9ef220c328938d7ca077f14c4ea85c5649b765e